File: scheduler/scheduler.js
#!/usr/bin/env node
/**
 * LiteQ 스케줄러 — 993문항 문제 은행에서 자동 세트 조립
 *
 * Usage:
 *   node scheduler.js --stats
 *   node scheduler.js --check-shortage
 *   node scheduler.js --day mon
 *   node scheduler.js --day mon --save
 *   node scheduler.js --week 2026-03-23 --save
 */

const fs = require("fs");
const path = require("path");

// ── Question Bank ──
class QuestionBank {
  constructor() { this.questions = []; this._index = {}; }

  load(basePath) {
    this.questions = []; this._index = {};
    for (const dir of fs.readdirSync(basePath)) {
      const dp = path.join(basePath, dir);
      if (!fs.statSync(dp).isDirectory()) continue;
      for (const f of fs.readdirSync(dp)) {
        if (!f.endsWith("_questions.json")) continue;
        const data = JSON.parse(fs.readFileSync(path.join(dp, f), "utf8"));
        for (const q of data) { this.questions.push(q); this._index[q.qid] = q; }
      }
    }
    console.log(`Loaded ${this.questions.length} questions`);
  }

  filter({ subject, unit, diff, mode, excludeQids, excludeRecentDays } = {}) {
    return this.questions.filter(q => {
      if (subject && q.subject !== subject) return false;
      if (unit && !q.unit?.includes(unit)) return false;
      if (diff != null) {
        const diffs = Array.isArray(diff) ? diff : [diff];
        if (!diffs.includes(q.diff)) return false;
      }
      if (mode && q.mode !== mode) return false;
      if (excludeQids && excludeQids.has(q.qid)) return false;
      if (excludeRecentDays && q.last_used) {
        const days = (Date.now() - new Date(q.last_used).getTime()) / 86400000;
        if (days < excludeRecentDays) return false;
      }
      return q.status === "approved";
    });
  }

  getByTags(tags, { subject, excludeQids } = {}) {
    const tagSet = new Set(tags);
    return this.questions
      .filter(q => {
        if (subject && q.subject !== subject) return false;
        if (excludeQids && excludeQids.has(q.qid)) return false;
        return (q.tags || []).some(t => tagSet.has(t));
      })
      .sort((a, b) => (a.used_count || 0) - (b.used_count || 0));
  }

  markUsed(qid) {
    const q = this._index[qid];
    if (q) { q.used_count = (q.used_count || 0) + 1; q.last_used = new Date().toISOString().slice(0, 10); }
  }

  getStats() {
    const s = { total: this.questions.length, bySubject: {}, byDiff: {} };
    for (const q of this.questions) {
      s.bySubject[q.subject] = (s.bySubject[q.subject] || 0) + 1;
      s.byDiff[q.diff] = (s.byDiff[q.diff] || 0) + 1;
    }
    return s;
  }
}

// ── Shuffle ──
function shuffle(arr) {
  const a = [...arr];
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

// ── Assembler ──
function assembleDailyOrMock(bank, config) {
  const usedQids = new Set();
  const topicCounts = {};
  const topicMax = config.topic_max_repeat || 99;
  const selected = [];

  for (const comp of config.composition || []) {
    const diffs = Array.isArray(comp.diff) ? comp.diff : [comp.diff];
    let candidates = bank.filter({
      subject: config.subject,
      diff: diffs,
      excludeQids: usedQids,
      excludeRecentDays: config.exclude_recent_days,
    });
    candidates = shuffle(candidates).sort((a, b) => (a.used_count || 0) - (b.used_count || 0));

    const modeMix = comp.mode_mix || {};
    const modeCounts = {};
    let picked = 0;

    for (const q of candidates) {
      if (picked >= comp.count) break;
      if (usedQids.has(q.qid)) continue;
      const topic = q.topic || "";
      if (topic && (topicCounts[topic] || 0) >= topicMax) continue;

      if (Object.keys(modeMix).length > 0) {
        const m = q.mode || "choice";
        if ((modeCounts[m] || 0) >= (modeMix[m] || 0)) continue;
        modeCounts[m] = (modeCounts[m] || 0) + 1;
      }

      selected.push(q);
      usedQids.add(q.qid);
      topicCounts[topic] = (topicCounts[topic] || 0) + 1;
      picked++;
    }

    // Fill remaining if mode_mix was too strict
    if (picked < comp.count) {
      for (const q of candidates) {
        if (picked >= comp.count) break;
        if (usedQids.has(q.qid)) continue;
        selected.push(q);
        usedQids.add(q.qid);
        picked++;
      }
    }
  }

  if (config.sort_by_difficulty) {
    selected.sort((a, b) => (a.diff || 0) - (b.diff || 0));
  }

  return selected;
}

function assembleWeekly(bank, config, wrongTags = []) {
  const total = config.total;
  const wrongRatio = config.wrong_tag_ratio || 0.6;
  const usedQids = new Set();
  const selected = [];

  if (wrongTags.length > 0) {
    const wrongCount = Math.round(total * wrongRatio);
    const wrongCandidates = bank.getByTags(wrongTags, { subject: config.subject, excludeQids: usedQids });
    for (const q of wrongCandidates.slice(0, wrongCount)) {
      selected.push(q); usedQids.add(q.qid);
    }
  }

  const remaining = total - selected.length;
  if (remaining > 0) {
    const pool = shuffle(bank.filter({
      subject: config.subject, excludeQids: usedQids,
      excludeRecentDays: config.exclude_recent_days,
    }));
    for (const q of pool.slice(0, remaining)) {
      selected.push(q); usedQids.add(q.qid);
    }
  }

  if (config.sort_by_difficulty) selected.sort((a, b) => (a.diff || 0) - (b.diff || 0));
  return selected;
}

// ── Validator ──
function validateSet(questions, config) {
  const errors = [];
  if (questions.length !== config.total) errors.push(`문항 수: ${questions.length}/${config.total}`);

  for (const comp of config.composition || []) {
    const diffs = Array.isArray(comp.diff) ? comp.diff : [comp.diff];
    const actual = questions.filter(q => diffs.includes(q.diff)).length;
    if (actual !== comp.count) errors.push(`diff=${comp.diff}: ${actual}/${comp.count}`);
  }

  const choiceQs = questions.filter(q => q.mode === "choice");
  if (choiceQs.length >= 5) {
    const dist = {};
    choiceQs.forEach(q => { dist[q.ans_index] = (dist[q.ans_index] || 0) + 1; });
    for (let i = 0; i < 5; i++) {
      if ((dist[i] || 0) / choiceQs.length > 0.35) errors.push(`정답 ${i+1}번 편중`);
    }
  }

  return { valid: errors.length === 0, errors };
}

// ── Configs ──
const CONFIGS = {
  MATH_DAILY: { subject: "수학", total: 7, time_limit: null, est_time: "20분", composition: [
    { diff: 1, pts: 2, count: 2, mode_mix: { choice: 1, short: 1 } },
    { diff: 2, pts: 3, count: 4, mode_mix: { choice: 2, short: 2 } },
    { diff: 3, pts: 4, count: 1, mode_mix: { short: 1 } },
  ], topic_max_repeat: 2, exclude_recent_days: 7 },

  MATH_MOCK: { subject: "수학", total: 18, time_limit: 60, est_time: "60분", composition: [
    { diff: 1, pts: 2, count: 2 },
    { diff: 2, pts: 3, count: 9 },
    { diff: [3, 4], pts: 4, count: 7 },
  ], sort_by_difficulty: true, exclude_recent_days: 14, killer_count: { min: 2, max: 3 } },

  KR_DAILY: { subject: "국어", total: 6, time_limit: null, est_time: "15분", composition: [
    { diff: 1, count: 2 }, { diff: 2, count: 2 }, { diff: [3, 4], count: 2 },
  ], exclude_recent_days: 7 },

  KR_MOCK: { subject: "국어", total: 15, time_limit: 30, est_time: "30분", composition: [
    { diff: 1, count: 4 }, { diff: 2, count: 6 }, { diff: [3, 4], count: 5 },
  ], sort_by_difficulty: true, exclude_recent_days: 14 },

  EN_DAILY: { subject: "영어", total: 14, time_limit: null, est_time: "25분", composition: [
    { diff: 1, count: 5 }, { diff: 2, count: 5 }, { diff: [3, 4], count: 4 },
  ], exclude_recent_days: 7 },

  EN_MOCK: { subject: "영어", total: 14, time_limit: 25, est_time: "25분", composition: [
    { diff: 1, count: 5 }, { diff: 2, count: 5 }, { diff: [3, 4], count: 4 },
  ], sort_by_difficulty: true, exclude_recent_days: 14 },

  ETHICS_DAILY: { subject: "생활과윤리", total: 10, time_limit: null, est_time: "20분", composition: [
    { diff: 1, count: 3 }, { diff: 2, count: 4 }, { diff: [3, 4], count: 3 },
  ], exclude_recent_days: 7 },

  EARTH_DAILY: { subject: "지구과학1", total: 10, time_limit: null, est_time: "20분", composition: [
    { diff: 1, count: 3 }, { diff: 2, count: 4 }, { diff: [3, 4], count: 3 },
  ], exclude_recent_days: 7 },
};

// ── Schedule ──
const SCHEDULE = {
  mon: [
    { type: "daily", config: CONFIGS.KR_DAILY },
    { type: "daily", config: CONFIGS.MATH_DAILY },
    { type: "daily", config: CONFIGS.EARTH_DAILY },
  ],
  tue: [
    { type: "daily", config: CONFIGS.EN_DAILY },
    { type: "daily", config: CONFIGS.MATH_DAILY },
    { type: "daily", config: CONFIGS.ETHICS_DAILY },
  ],
  wed: [
    { type: "daily", config: CONFIGS.KR_DAILY },
    { type: "mock", config: CONFIGS.MATH_MOCK },
  ],
  thu: [
    { type: "daily", config: CONFIGS.EN_DAILY },
    { type: "daily", config: CONFIGS.MATH_DAILY },
    { type: "daily", config: CONFIGS.EARTH_DAILY },
  ],
  fri: [
    { type: "daily", config: CONFIGS.KR_DAILY },
    { type: "daily", config: CONFIGS.MATH_DAILY },
    { type: "daily", config: CONFIGS.ETHICS_DAILY },
  ],
  sat: [
    { type: "mock", config: CONFIGS.KR_MOCK },
    { type: "mock", config: CONFIGS.EN_MOCK },
  ],
  sun: null,
};

const DAY_NAMES = ["mon", "tue", "wed", "thu", "fri", "sat", "sun"];

// ── Generate Day ──
function generateDay(bank, dayName, dateStr) {
  const schedule = SCHEDULE[dayName];
  if (!schedule) { console.log(`  ${dayName}: 휴식일`); return null; }

  const sets = [];
  let totalQ = 0;

  for (const entry of schedule) {
    const config = entry.config;
    const questions = entry.type === "weekly"
      ? assembleWeekly(bank, config, [])
      : assembleDailyOrMock(bank, config);

    const result = validateSet(questions, config);
    const mark = result.valid ? "✓" : "⚠";
    if (!result.valid) result.errors.forEach(e => console.log(`    ${e}`));

    questions.forEach(q => bank.markUsed(q.qid));
    sets.push({ subject: config.subject, type: entry.type, total: questions.length,
      est_time: config.est_time, time_limit: config.time_limit, questions });
    totalQ += questions.length;
    console.log(`  ${mark} ${config.subject} ${entry.type}: ${questions.length}/${config.total}`);
  }

  return {
    meta: { date: dateStr, day: dayName, total_questions: totalQ,
      sets: sets.map(s => ({ subject: s.subject, type: s.type, total: s.total, est_time: s.est_time })) },
    sets: sets.map(s => ({ subject: s.subject, type: s.type, questions: s.questions })),
  };
}

// ── CLI ──
const args = process.argv.slice(2);
const flags = {};
for (let i = 0; i < args.length; i++) {
  if (args[i].startsWith("--")) {
    const key = args[i].slice(2);
    flags[key] = (i + 1 < args.length && !args[i + 1]?.startsWith("--")) ? args[++i] : true;
  }
}

const bank = new QuestionBank();
bank.load(path.join(__dirname, "..", "data", "question_bank"));

if (flags.stats) {
  const s = bank.getStats();
  console.log("\n=== LiteQ 문제 은행 현황 ===");
  Object.entries(s.bySubject).sort((a, b) => b[1] - a[1]).forEach(([subj, cnt]) => {
    const qs = bank.filter({ subject: subj });
    const diffs = {};
    qs.forEach(q => { diffs[q.diff] = (diffs[q.diff] || 0) + 1; });
    const d = Object.entries(diffs).sort((a, b) => a[0] - b[0]).map(([k, v]) => `d${k}:${v}`).join(" ");
    console.log(`  ${subj.padEnd(12)} ${String(cnt).padStart(4)}문항  (${d})`);
  });
  console.log(`  ${"─".repeat(40)}`);
  console.log(`  ${"총계".padEnd(12)} ${String(s.total).padStart(4)}문항`);
  process.exit(0);
}

if (flags["check-shortage"]) {
  console.log("\n=== 부족 영역 (4주 운영 기준) ===");
  const reqs = { "수학": 284, "국어": 120, "영어": 200, "생활과윤리": 80, "지구과학1": 80 };
  for (const [subj, needed] of Object.entries(reqs)) {
    const total = bank.filter({ subject: subj }).length;
    if (total < needed) console.log(`  ⚠ ${subj}: ${total}문항 / 필요 ${needed} → ${needed - total}문항 부족`);
    else console.log(`  ✓ ${subj}: ${total}문항 (충분)`);
  }
  process.exit(0);
}

if (flags.day) {
  const dayName = flags.day;
  const dateStr = flags.date || new Date().toISOString().slice(0, 10);
  console.log(`\n=== ${dateStr} (${dayName}) 세트 생성 ===`);
  const output = generateDay(bank, dayName, dateStr);
  if (output && flags.save) {
    const outDir = path.join(__dirname, "output");
    if (!fs.existsSync(outDir)) fs.mkdirSync(outDir, { recursive: true });
    const fname = `${dateStr}_${dayName}.json`;
    fs.writeFileSync(path.join(outDir, fname), JSON.stringify(output, null, 2), "utf8");
    console.log(`\n저장: output/${fname}`);
  }
  if (output) console.log(`\n총 ${output.meta.total_questions}문항 생성`);
  process.exit(0);
}

if (flags.week) {
  const start = new Date(flags.week);
  console.log(`\n=== ${flags.week} 주간 세트 생성 ===\n`);
  let weekTotal = 0;
  for (let i = 0; i < 7; i++) {
    const dt = new Date(start); dt.setDate(dt.getDate() + i);
    const dayName = DAY_NAMES[dt.getDay() === 0 ? 6 : dt.getDay() - 1];
    const dateStr = dt.toISOString().slice(0, 10);
    console.log(`--- ${dateStr} (${dayName}) ---`);
    const output = generateDay(bank, dayName, dateStr);
    if (output) {
      weekTotal += output.meta.total_questions;
      if (flags.save) {
        const outDir = path.join(__dirname, "output");
        if (!fs.existsSync(outDir)) fs.mkdirSync(outDir, { recursive: true });
        fs.writeFileSync(path.join(outDir, `${dateStr}_${dayName}.json`), JSON.stringify(output, null, 2), "utf8");
      }
    }
    console.log();
  }
  console.log(`주간 총 ${weekTotal}문항 생성`);
  process.exit(0);
}

// Default
const s = bank.getStats();
console.log(`\n문제 은행: ${s.total}문항. --stats, --day, --week 옵션을 사용하세요.`);

File: scheduler/scheduler.py
#!/usr/bin/env python3
"""
LiteQ 스케줄러 — 993문항 문제 은행에서 자동 세트 조립

Usage:
  python scheduler.py --day mon
  python scheduler.py --date 2026-03-23
  python scheduler.py --week 2026-03-23
  python scheduler.py --subject 수학 --type daily
  python scheduler.py --subject 수학 --type weekly --wrong-tags "수열,점화식"
  python scheduler.py --stats
  python scheduler.py --check-shortage
"""

import argparse
import json
import os
import sys
from datetime import datetime, timedelta

# Add scheduler dir to path
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from question_bank import QuestionBank
from assembler import SetAssembler
from schedule import WEEKLY_SCHEDULE, DAY_NAMES
from configs import *


def get_bank_path():
    return os.path.join(os.path.dirname(__file__), "..", "data", "question_bank")


def get_output_path():
    return os.path.join(os.path.dirname(__file__), "output")


def load_bank():
    bank = QuestionBank()
    bank.load(get_bank_path())
    return bank


def generate_day(bank, day_name, date_str=None, wrong_tags=None):
    """특정 요일의 세트 생성"""
    schedule = WEEKLY_SCHEDULE.get(day_name)
    if schedule is None:
        print(f"{day_name}: 휴식일 (세트 없음)")
        return None

    assembler = SetAssembler(bank)
    sets = []
    total_q = 0

    for entry in schedule:
        set_type = entry["type"]
        config = entry["config"]

        if set_type == "daily":
            questions = assembler.assemble_daily(config)
        elif set_type == "mock":
            questions = assembler.assemble_mock(config)
        elif set_type == "weekly":
            questions = assembler.assemble_weekly(config, wrong_tags or [])
        else:
            continue

        # Validate
        result = assembler.validate(questions, config)
        status = "✓" if result["valid"] else "⚠"
        if not result["valid"]:
            for err in result["errors"]:
                print(f"  {config['subject']} {err}")

        # Mark used
        for q in questions:
            bank.mark_used(q["qid"])

        sets.append({
            "subject": config["subject"],
            "type": set_type,
            "total": len(questions),
            "est_time": config.get("est_time", ""),
            "time_limit": config.get("time_limit"),
            "valid": result["valid"],
            "questions": questions,
        })
        total_q += len(questions)
        print(f"  {status} {config['subject']} {set_type}: {len(questions)}/{config['total']}문항")

    output = {
        "meta": {
            "date": date_str or datetime.now().strftime("%Y-%m-%d"),
            "day": day_name,
            "sets": [{
                "subject": s["subject"],
                "type": s["type"],
                "total": s["total"],
                "est_time": s["est_time"],
                "time_limit": s["time_limit"],
            } for s in sets],
            "total_questions": total_q,
        },
        "sets": [{
            "subject": s["subject"],
            "type": s["type"],
            "questions": s["questions"],
        } for s in sets],
    }

    return output


def print_stats(bank):
    """문제 은행 현황"""
    stats = bank.get_stats()
    print("\n=== LiteQ 문제 은행 현황 ===")
    for subj, count in sorted(stats["by_subject"].items(), key=lambda x: -x[1]):
        # Get diff breakdown
        subj_qs = bank.filter(subject=subj)
        diffs = {}
        for q in subj_qs:
            d = q.get("diff", 0)
            diffs[d] = diffs.get(d, 0) + 1
        diff_str = " ".join(f"d{k}:{v}" for k, v in sorted(diffs.items()))
        print(f"  {subj:12s} {count:4d}문항  ({diff_str})")
    print(f"  {'─' * 40}")
    print(f"  {'총계':12s} {stats['total']:4d}문항")


def check_shortage(bank):
    """부족 영역 확인 (4주 운영 기준)"""
    print("\n=== 부족 영역 (4주 운영 기준) ===")

    requirements = {
        "수학": {"total_needed": 284, "diff4_needed": 20},
        "국어": {"total_needed": 120, "diff4_needed": 0},
        "영어": {"total_needed": 200, "diff4_needed": 0},
        "생활과윤리": {"total_needed": 80, "diff4_needed": 0},
        "지구과학1": {"total_needed": 80, "diff4_needed": 0},
    }

    for subj, req in requirements.items():
        qs = bank.filter(subject=subj)
        total = len(qs)
        needed = req["total_needed"]
        if total < needed:
            print(f"  ⚠ {subj}: {total}문항 / 필요 {needed} → {needed - total}문항 부족")
        else:
            print(f"  ✓ {subj}: {total}문항 (충분)")


def main():
    parser = argparse.ArgumentParser(description="LiteQ 스케줄러")
    parser.add_argument("--day", help="요일 (mon~sun)")
    parser.add_argument("--date", help="날짜 (YYYY-MM-DD)")
    parser.add_argument("--week", help="주 시작일 (YYYY-MM-DD)")
    parser.add_argument("--subject", help="과목")
    parser.add_argument("--type", help="유형 (daily/mock/weekly)")
    parser.add_argument("--wrong-tags", help="오답 태그 (쉼표 구분)")
    parser.add_argument("--stats", action="store_true", help="문제 은행 현황")
    parser.add_argument("--check-shortage", action="store_true", help="부족 영역 확인")
    parser.add_argument("--save", action="store_true", help="결과를 파일로 저장")
    args = parser.parse_args()

    bank = load_bank()

    if args.stats:
        print_stats(bank)
        return

    if args.check_shortage:
        check_shortage(bank)
        return

    # Single day
    if args.day or args.date:
        day_name = args.day
        date_str = args.date

        if date_str and not day_name:
            dt = datetime.strptime(date_str, "%Y-%m-%d")
            day_name = DAY_NAMES[dt.weekday()]

        if not day_name:
            day_name = DAY_NAMES[datetime.now().weekday()]
            date_str = datetime.now().strftime("%Y-%m-%d")

        print(f"\n=== {date_str or 'today'} ({day_name}) 세트 생성 ===")
        output = generate_day(bank, day_name, date_str,
                              wrong_tags=args.wrong_tags.split(",") if args.wrong_tags else None)

        if output and args.save:
            out_dir = get_output_path()
            os.makedirs(out_dir, exist_ok=True)
            fname = f"{date_str or day_name}.json"
            with open(os.path.join(out_dir, fname), "w", encoding="utf-8") as f:
                json.dump(output, f, ensure_ascii=False, indent=2)
            print(f"\n저장: output/{fname}")

        if output:
            total = output["meta"]["total_questions"]
            print(f"\n총 {total}문항 생성 완료")

        return

    # Full week
    if args.week:
        start = datetime.strptime(args.week, "%Y-%m-%d")
        print(f"\n=== {args.week} 주간 세트 생성 ===\n")
        week_total = 0

        for i in range(7):
            dt = start + timedelta(days=i)
            day_name = DAY_NAMES[dt.weekday()]
            date_str = dt.strftime("%Y-%m-%d")
            print(f"--- {date_str} ({day_name}) ---")
            output = generate_day(bank, day_name, date_str)
            if output:
                week_total += output["meta"]["total_questions"]
                if args.save:
                    out_dir = get_output_path()
                    os.makedirs(out_dir, exist_ok=True)
                    fname = f"{date_str}_{day_name}.json"
                    with open(os.path.join(out_dir, fname), "w", encoding="utf-8") as f:
                        json.dump(output, f, ensure_ascii=False, indent=2)
            print()

        print(f"주간 총 {week_total}문항 생성")
        return

    # Default: show stats
    print_stats(bank)
    check_shortage(bank)


if __name__ == "__main__":
    main()
